libtransmission/file-posix.c

```diff
@@ -430,6 +452,124 @@ bool tr_sys_path_rename(char const* src_path, char const* dst_path, tr_error** e
     return ret;
 }
 
+/* We try to do a fast (in-kernel) copy using a variety of non-portable system
+ * calls. If the current implementation does not support in-kernel copying, we
+ * use a user-space fallback instead. */
+bool tr_sys_path_copy(char const* src_path, char const* dst_path, tr_error** error)
+{
+    TR_ASSERT(src_path != NULL);
+    TR_ASSERT(dst_path != NULL);
+
+#if defined(USE_COPYFILE)
+    if (copyfile(src_path, dst_path, NULL, COPYFILE_CLONE | COPYFILE_ALL) < 0)
+    {
+        set_system_error(error, errno);
+        return false;
+    }
+
+    return true;
+
+#else /* USE_COPYFILE */
+
+    /* Other OSes require us to copy between file descriptors, so open them. */
+    tr_sys_file_t in = tr_sys_file_open(src_path, TR_SYS_FILE_READ | TR_SYS_FILE_SEQUENTIAL, 0, error);
+    if (in == TR_BAD_SYS_FILE)
+    {
+        tr_error_prefix(error, "Unable to open source file: ");
+        return false;
+    }
+
+    tr_sys_path_info info;
+    if (!tr_sys_file_get_info(in, &info, error))
+    {
+        tr_error_prefix(error, "Unable to get information on source file: ");
+        tr_sys_file_close(in, NULL);
+        return false;
+    }
+
+    tr_sys_file_t out = tr_sys_file_open(dst_path, TR_SYS_FILE_WRITE | TR_SYS_FILE_CREATE | TR_SYS_FILE_TRUNCATE, 0666, error);
+    if (out == TR_BAD_SYS_FILE)
+    {
+        tr_error_prefix(error, "Unable to open destination file: ");
+        tr_sys_file_close(in, NULL);
+        return false;
+    }
+
+    size_t file_size = info.size;
+
+#if defined(USE_COPY_FILE_RANGE) || defined(USE_SENDFILE64)
+
+    while (file_size > 0)
+    {
+        size_t const chunk_size = MIN(file_size, SSIZE_MAX);
+        ssize_t const copied =
+#ifdef USE_COPY_FILE_RANGE
+            copy_file_range(in, NULL, out, NULL, chunk_size, 0);
+#elif defined(USE_SENDFILE64)
+            sendfile64(out, in, NULL, chunk_size);
+#else
+#error
+#endif
+        TR_ASSERT(copied == -1 || copied >= 0); /* -1 for error; some non-negative value otherwise. */
+
+        if (copied == -1)
+        {
+            set_system_error(error, errno);
+            break;
+        }
+
+        TR_ASSERT(copied >= 0 && ((uint64_t)copied) <= file_size);
```

Comment:
I'd say it should not only be less or equal to `file_size` but also `chunk_size`, and since the latter is never greater than the former maybe we want to use it instead to make the assertion stricter. But I get why it's in its current form right now, so won't insist.